### ISSUE TITLE
[Linux] screen reader often stops working for Flutter apps

### ISSUE THREAD

**Issue by jpnurmi** [flutter/flutter#101513]
## Steps to Reproduce

1. Execute `flutter create -t skeleton foo`
2. Run the Orca screen reader
  a. Execute `tail -F orca.log` in one xterm
  b. Execute `orca -r --debug-file=orca.log` in another xterm 
3. Execute `flutter run`
4. Tab-navigate around the UI
5. Restart the app if Orca reads out widgets
6. Repeat steps 3-5 (until Orca stops reading the Flutter widgets)

**Expected results:** 

Orca continues to work regardless of restarting the Flutter app.

**Actual results:** 

Orca eventually stops working. According to `orca.log` the accessibility node tree breaks. You should see events similar to this:
```
vvvvv PROCESS OBJECT EVENT object:state-changed:focused vvvvv
15:41:10.882466 - OBJECT EVENT: object:state-changed:focused (1, 0, 0)
                                app.name='a11y_bug' name='' role='push button'
                                description=''
                                states='enabled focusable focused showing visible'
                                relations=''
                                interfaces='Accessible Action Collection Component Text'
                                attributes=''
                                path=[-1, 0, 0, 0, 0, -1, -1, -1, -1]

15:41:10.883095 - EVENT MANAGER: Getting script for [application | a11y_bug] (check: True)
15:41:10.885399 - EVENT MANAGER: Script is <orca.scripts.toolkits.gtk.script.Script object at 0x7f867d3e5a50>
15:41:10.885450 - EVENT MANAGER: Change active script: False (The script for this event is already active.)
15:41:10.887025 - INFO: Looking at [frame | a11y_bug] from [application | a11y_bug] /home/jpnurmi/Temp/a11y_bug/build/linux/x64/debug/bundle/a11y_bug 
15:41:10.887099 - INFO: [frame | a11y_bug] can be active window
15:41:10.887121 - ORCA: Active script is: <orca.scripts.toolkits.gtk.script.Script object at 0x7f867d3e5a50>
15:41:10.887200 - ZOMBIE: [push button | ]'s index is -1
15:41:10.887241 - ERROR: New locusOfFocus ([push button | ]) is zombie. Not updating.
15:41:10.887288 - EVENT MANAGER: locusOfFocus: [panel | ] activeScript: <orca.scripts.toolkits.gtk.script.Script object at 0x7f867d3e5a50>
TOTAL PROCESSING TIME: 0.0048
^^^^^ PROCESS OBJECT EVENT object:state-changed:focused ^^^^^
```

<details>
<summary>Code sample</summary>

```dart
```

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel master, 2.13.0-0.0.pre.424, on Ubuntu Jammy Jellyfish (development branch) 5.15.0-25-generic, locale en_US.UTF-8)
    • Flutter version 2.13.0-0.0.pre.424 at /home/jpnurmi/Flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 31668eef40 (7 hours ago), 2022-04-07 09:51:08 +0300
    • Engine revision 541b636563
    • Dart version 2.17.0 (build 2.17.0-266.0.dev)
    • DevTools version 2.12.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/jpnurmi/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_SDK_ROOT = /home/jpnurmi/Android/Sdk
    • Java binary at: /home/jpnurmi/Android/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/jpnurmi/Android/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension version 3.38.1

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Ubuntu Jammy Jellyfish (development branch) 5.15.0-25-generic

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by jpnurmi** [comment]
Now that I think of it, there's probably a nice and reliable way to reproduce the issue by forcing accessibility tree changes.

**Comment by jpnurmi** [comment]
The bug is that [`fl_view_accessible_handle_update_semantics_node()`](https://github.com/flutter/engine/blob/main/shell/platform/linux/fl_view_accessible.cc#L73) notifies state changes first, and then updates the node tree. Presumably, it should be done the other way around to ensure that the accessibility node tree is up-to-date so that the object's location in the tree is known when notifying state changes.

Here's the respective piece of code in Orca: https://gitlab.gnome.org/GNOME/orca/-/blob/master/src/orca/script_utilities.py#L5041

**Comment by maheshmnj** [comment]
Hi @jpnurmi, thanks for filing the issue. I am able to reproduce the issue on stable and the master channel. The orca doesn't read the widgets most of the time on my machine. I do the similar logs as in the original post.
  
logs can be found here [orca.zip](https://github.com/flutter/flutter/files/8450002/orca.zip)

<details>
<summary>flutter doctor -v (Linux)</summary>

```
[✓] Flutter (Channel stable, 2.10.4, on Ubuntu 20.04.4 LTS 5.13.0-30-generic, locale en_GB.UTF-8)
    • Flutter version 2.10.4 at /home/mahesh/Development/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c860cba910 (12 days ago), 2022-03-25 00:23:12 -0500
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin version 63.2.1
    • Dart plugin version 203.8452
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.64.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.38.1

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.4 LTS 5.13.0-30-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 98.0.4758.102

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
```
[✓] Flutter (Channel master, 2.13.0-0.0.pre.392, on Ubuntu 20.04.4 LTS 5.13.0-30-generic, locale en_GB.UTF-8)
    • Flutter version 2.13.0-0.0.pre.392 at /home/mahesh/Development/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 120b3deb18 (20 minutes ago), 2022-04-06 10:59:37 +0200
    • Engine revision 45bedbe841
    • Dart version 2.17.0 (build 2.17.0-266.0.dev)
    • DevTools version 2.12.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin version 63.2.1
    • Dart plugin version 203.8452
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.64.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.38.1

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.4 LTS 5.13.0-30-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 98.0.4758.102

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by jpnurmi** [comment]
Hi @maheshmnj, that's great to hear! I was worried it would be too random to reproduce. Thanks.

**Comment by cbracken** [comment]
@jpnurmi do you think your fix for this issue might have fixed https://github.com/flutter/flutter/issues/95230? I'm wondering if they might be the same underlying issue.

**Comment by jpnurmi** [comment]
@cbracken It sounds a lot like the same issue. I'm not familiar with NVDA but I can take a look tomorrow if I could get it up and running.
